pkg/cgroups/cgroupsv2/runtime.go

```diff
@@ -0,0 +1,147 @@
+package cgroupsv2
+
+import (
+	"context"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"log/slog"
+	"os"
+	"os/exec"
+	"syscall"
+	"time"
+
+	"github.com/google/uuid"
+	jobv1 "github.com/kralicky/jobserver/pkg/apis/job/v1"
+	"github.com/kralicky/jobserver/pkg/cgroups"
+	"github.com/kralicky/jobserver/pkg/jobs"
+	"github.com/kralicky/jobserver/pkg/util"
+)
+
+const gracePeriod = 10 * time.Second
+
+type v2Runtime struct {
+	mgr *cgroupManager
+}
+
+func NewRuntime() (jobs.Runtime, error) {
+	mgr, err := newCgroupManager()
+	if err != nil {
+		return nil, fmt.Errorf("failed to setup jobserver cgroup: %w", err)
+	}
+
+	return &v2Runtime{
+		mgr: mgr,
+	}, nil
+}
+
+// Execute implements jobs.Runtime.
+func (l *v2Runtime) Execute(ctx context.Context, spec *jobv1.JobSpec) (jobs.Process, error) {
+	cmdSpec := spec.GetCommand()
+
+	// generate a uuid for the process, but encode it in the raw hex format.
+	// this makes it slightly easier to use from the command line, since many
+	// terminals treat '-' as a word separator, making it difficult to copy
+	// the whole string using a double-click.
+	u := uuid.New()
+	id := hex.EncodeToString(u[:])
+
+	cmd := exec.CommandContext(ctx, cmdSpec.GetCommand(), cmdSpec.GetArgs()...)
+	cmd.Env = append(os.Environ(), cmdSpec.GetEnv()...)
+
+	streamBuf := util.NewStreamBuffer()
+	done := make(chan struct{})
+
+	cmd.Stdout = streamBuf
+	cmd.Stderr = streamBuf
+	cmd.Stdin = nil
+	cmd.Cancel = func() error {
+		lg := slog.With("id", id)
+		lg.Debug("context canceled; attempting graceful shutdown")
+		streamBuf.Close() // NB: leaving this open will cause cmd.Wait to hang
+		start := time.Now()
+		go func() {
+			timeout := time.NewTimer(gracePeriod)
+			defer timeout.Stop()
+			select {
+			case <-timeout.C:
+				lg.Warn("process did did not exit within grace period, sending SIGKILL")
+				if err := cmd.Process.Signal(syscall.SIGKILL); err != nil {
+					lg.Error("failed to send SIGKILL", "error", err)
+				}
+			case <-done:
+				lg.Debug("process exited within grace period", "took", time.Since(start))
+			}
+		}()
```

Comment:
Don't you get this exact behavior by setting `WaitDelay` on the `Cmd`?

Reply:
Yep, WaitDelay seems to work. :+1: